tests/math/test_functions.py

```diff
@@ -257,6 +377,104 @@ def test_convert_scalar(self, t_like):
+        res = fn.dot(t1, t2)
+        assert fn.allequal(res, 14)
+
+    matrix_vector_product_data = [
+        [[[1, 2], [3, 4]], [6, 7]],
+        [np.array([[1, 2], [3, 4]]), np.array([6, 7])],
+        [torch.tensor([[1, 2], [3, 4]]), onp.array([6, 7])],
+        [torch.tensor([[1, 2], [3, 4]]), torch.tensor([6, 7])],
+        [tf.Variable([[1, 2], [3, 4]]), onp.array([6, 7])],
+        [tf.constant([[1, 2], [3, 4]]), onp.array([6, 7])],
+        [tf.Variable([[1, 2], [3, 4]]), tf.Variable([6, 7])],
+    ]
+
+    @pytest.mark.parametrize("t1, t2", matrix_vector_product_data)
+    def test_matrix_vector_product(self, t1, t2):
+        """Test that the matrix-vector dot product of two vectors results in a vector"""
+        res = fn.dot(t1, t2)
+        assert fn.allequal(res, [20, 46])
+
+    @pytest.mark.parametrize("t1, t2", matrix_vector_product_data)
+    def test_vector_matrix_product(self, t1, t2):
+        """Test that the vector-matrix dot product of two vectors results in a vector"""
+        res = fn.dot(t2, t1)
+        assert fn.allequal(res, [27, 40])
+
+    @pytest.mark.parametrize("t1, t2", matrix_vector_product_data)
+    def test_matrix_matrix_product(self, t1, t2):
+        """Test that the matrix-matrix dot product of two vectors results in a matrix"""
+        res = fn.dot(t1, t1)
+        assert fn.allequal(res, np.array([[7, 10], [15, 22]]))
+
+    multidim_product_data = [
+        [np.array([[[1, 2], [3, 4], [-1, 1]], [[5, 6], [0, -1], [2, 1]]]), np.array([[[1, 1], [3, 3]], [[3, 1], [3, 2]]])],
+        [torch.tensor([[[1, 2], [3, 4], [-1, 1]], [[5, 6], [0, -1], [2, 1]]]), onp.array([[[1, 1], [3, 3]], [[3, 1], [3, 2]]])],
+        [torch.tensor([[[1, 2], [3, 4], [-1, 1]], [[5, 6], [0, -1], [2, 1]]]), torch.tensor([[[1, 1], [3, 3]], [[3, 1], [3, 2]]])],
+        [onp.array([[[1, 2], [3, 4], [-1, 1]], [[5, 6], [0, -1], [2, 1]]]), tf.Variable([[[1, 1], [3, 3]], [[3, 1], [3, 2]]])],
+        [tf.constant([[[1, 2], [3, 4], [-1, 1]], [[5, 6], [0, -1], [2, 1]]]), onp.array([[[1, 1], [3, 3]], [[3, 1], [3, 2]]])],
+        [tf.Variable([[[1, 2], [3, 4], [-1, 1]], [[5, 6], [0, -1], [2, 1]]]), tf.constant([[[1, 1], [3, 3]], [[3, 1], [3, 2]]])],
+    ]
+
+    @pytest.mark.parametrize("t1, t2", multidim_product_data)
+    def test_multidimensional_product(self, t1, t2):
+        """Test that the multi-dimensional dot product reduces across the last dimension of the first
+        tensor, and the second-to-last dimension of the second tensor."""
+        res = fn.dot(t1, t2)
+        expected = np.array([[[[ 7,  7],
+              [ 9,  5]],
+
+             [[15, 15],
+              [21, 11]],
+
+             [[ 2,  2],
+              [ 0,  1]]],
+
+
+            [[[23, 23],
+              [33, 17]],
+
+             [[-3, -3],
+              [-3, -2]],
+
+             [[ 5,  5],
+              [ 9,  4]]]]
+        )
+        assert fn.allequal(res, expected)

```

Comment:
nice test data

Reply:
I tried to test as many edge cases as possible - you'd be surprised how differently torch/numpy/tensorflow treat something as simple as tensordot!